### PR TITLE
[Relay] change device annotation from post DFS to recursive

### PR DESCRIPTION
**Pull request by zhanghaohit**
This is related to #5840 and split from PR #5842 

Originally, device type is propagated based on the post DFS traversed graph, which may not be consistent if the argument order changes. In addition, it may handle some cases wrongly, e.g., the first residual block in Resnet50. The first few layers in Resnet50 are depicted in the following figure (top to bottom is in DFS order). Basically, we want to let all the layers run on FPGA device, except the first and last few layers. In the original device propagation algorithm, based on the post DFS order, the conv2d layers in grey will be propagated with `CPU` device type as we encounter `copy2` first, following which the three grey conv2d nodes are marked as the source device type of `copy2` (i.e., `CPU`), which is not correct.

Resnet50

By change the device annotation behaviour, we can support more complex graph structure.

### CODE REVIEW

**Comment by jroesch** [comment]
cc @mbrookhart

**Comment by zhanghaohit** [comment]
> Thanks @zhanghaohit for this PR, I second @mbrookhart's request to add a test case.

**Comment by zhanghaohit** [comment]
> Can you please add a test showing the problem/desired change in behavior?

Thanks @mbrookhart and @tmoreau89 for the suggestion. I've added a test, which would fail in the original code.

File: tests/python/relay/test_pass_annotation.py [hunk 1]
```diff
@@ -309,6 +309,76 @@ def test_visitor_annotation():
...
+        add = relay.add(log2, log10)
+        _add = relay.device_copy(add, ctx2, ctx1)
+        tan = relay.tan(_add)
+
+        func = run_opt_pass(tan, transform.InferType())
+        return func
+
+    annotated_expr = annotated()
+    expected_expr = expected()
+    assert tvm.ir.structural_equal(annotated_expr, expected_expr)
```
Comment (mbrookhart) [inline tests/python/relay/test_pass_annotation.py:367]: I'm curious that master passes this check, but fails on line 377. Why doesn't structural equal properly resolve the error in the device copy op?

Comment (zhanghaohit) [reply tests/python/relay/test_pass_annotation.py:367]: I think master fails on line 379, right? the device type of `log2` is not correctly marked.

Up to this line, `annotated_expr` and `expected_expr` are exactly the same. The `device_copy` op is inserted correctly. We haven't go through the device propagation yet.

Comment (mbrookhart) [reply tests/python/relay/test_pass_annotation.py:367]: :+1: You're right, I misread my first test.

**Comment by zhiics** [comment]
please trigger the ci again.

**Comment by zhanghaohit** [comment]
@tmoreau89 Could you help merge this PR? Thanks.

**Comment by tmoreau89** [comment]
Thank you @zhanghaohit , @mbrookhart, @junrushao1994, @zhiics the PR has been merged.

**Pull request merged**